### DISCUSSION TITLE
Newbie: retrieving in memory image data for use with PIL?

### DISCUSSION THREAD

**Discussion by felciano** [RhetTbull/osxphotos#1062] [General]
Hi folks --

I'm not sure if this is a bug (presumably a documentation one) or a feature request yet. I'm trying to manipulate an image from the OSX photos library in my script, using [pillow](https://pillow.readthedocs.io/en/stable/) to do the image editing. I cannot find the API call that gets me access to the binary image data in a way that could be passed to PIL. Is `osxphotos` just restricted to working with image metadata? Is there an alternate way to get at the image data itself that I could then pass on to pillow?

Thanks in advance for your help.

**Comment by RhetTbull** [comment]
Not a bug nor a documentation issue as editing images is not something osxphotos provides for.  However, you can use osxphotos to get access to the original image on disk using any of the [PhotoInfo.path](https://rhettbull.github.io/osxphotos/reference.html#osxphotos.PhotoInfo.path) properties (e..g `photo.path` returns the path to the original image, `photo.path_edited` returns the path the edited image (or `None` if not edited).  A couple things to keep in mind when doing this:

* If you have `Optimize Mac Storage` set in your Photos settings then the photo may not actually be on disk and `path` would return `None`
* You **should not** edit the file in the Photos library.  This could potentially corrupt your library. If you want to grab the original image, edit it in pillow, then save that edited image outside your Photos library, that's a supported use case but osxphotos *does not* provide a means to edit the file in your library and have those edits appear in Photos. With the exception of a few utilities like `osxphotos timewarp` and `osxphotos import`, osxphotos generally provides *read only* access to the library and associated metadata.
* Another option would be to export the photo to a temp file using [PhotoInfo.export()](https://rhettbull.github.io/osxphotos/reference.html#osxphotos.PhotoInfo.export).  There's an [example](https://github.com/RhetTbull/osxphotos/blob/main/examples/export_faces.py) in the examples folder that does this (and uses pillow to edit the image).  If you do have `Optimize Mac Storage` set, you can pass `use_photos_export=True` to `export()` to cause Photos to download the missing file from iCloud upon export.

**Comment by RhetTbull** [comment]
Also, you might find the [API_README](https://github.com/RhetTbull/osxphotos/blob/main/API_README.md) and the [osxphotos Python package reference](https://rhettbull.github.io/osxphotos/reference.html) helpful if you're using osxphotos in your own code.

**Comment by RhetTbull** [comment]
Hi @felciano I've moved this to the Discussion board. Did you get your code working? Cheers!
